x-pack/platform/plugins/shared/lens/public/visualizations/datatable/components/toolbar.tsx

```diff
@@ -157,3 +179,69 @@ export function DataTableToolbar(props: VisualizationToolbarProps<DatatableVisua
     </EuiFlexGroup>
   );
 }
+
+interface DensitySettingsProps {
+  dataGridDensity: DataGridDensity;
+  onChange: (density: DataGridDensity) => void;
+}
+
+const DENSITY_TOKENS = [
+  'euiDisplaySelector.densityLabel',
+  'euiDisplaySelector.labelCompact',
+  'euiDisplaySelector.labelNormal',
+  'euiDisplaySelector.labelExpanded',
+];
+
+const DENSITY_DEFAULTS = ['Density', 'Compact', 'Normal', 'Expanded'];
+
+const getDensityOptions = ({
+  labelCompact,
+  labelNormal,
+  labelExpanded,
+}: {
+  labelCompact: string;
+  labelNormal: string;
+  labelExpanded: string;
+}): EuiButtonGroupOptionProps[] => [
+  { id: DataGridDensity.COMPACT, label: labelCompact },
+  { id: DataGridDensity.NORMAL, label: labelNormal },
+  { id: DataGridDensity.EXPANDED, label: labelExpanded },
+];
+
+const DensitySettings: React.FC<DensitySettingsProps> = ({ dataGridDensity, onChange }) => {
+  const isValidDensity = (value: string): value is DataGridDensity => {
+    return Object.values(DataGridDensity).includes(value as DataGridDensity);
+  };
+
+  const setDensity = useCallback(
+    (density: string) => {
+      const _density = isValidDensity(density) ? density : DataGridDensity.NORMAL;
+      onChange(_density);
+    },
+    [onChange]
+  );
+
+  return (
+    <EuiI18n tokens={DENSITY_TOKENS} defaults={DENSITY_DEFAULTS}>
+      {([densityLabel, labelCompact, labelNormal, labelExpanded]: string[]) => {
+        const densityOptions = getDensityOptions({ labelCompact, labelNormal, labelExpanded });
+        return (
+          <EuiFormRow
+            label={densityLabel}
+            display="columnCompressed"
+            data-test-subj="lnsDensitySettings"
+          >
+            <EuiButtonGroup
+              legend={densityLabel}
+              buttonSize="compressed"
+              isFullWidth
+              options={densityOptions}
+              onChange={setDensity}
+              idSelected={dataGridDensity}
+            />
+          </EuiFormRow>
+        );
+      }}
+    </EuiI18n>
+  );
```

Comment:
I think this can be simplified without the `EuiI18n` component and with a leaner tokens declaration as in:
```ts
const densityLabel = i18n.translate('xpack.lens.table.densityLabel', {
    defaultMessage: 'Density',
  });

  return (
    <EuiFormRow label={densityLabel} display="columnCompressed" data-test-subj="lnsDensitySettings">
      <EuiButtonGroup
        legend={densityLabel}
        buttonSize="compressed"
        isFullWidth
        options={densityOptions}
        onChange={setDensity}
        idSelected={dataGridDensity}
      />
    </EuiFormRow>
  );
```

This has also the benefit of scoping the tokens for Lens.